delft/sequenceLabelling/preprocess.py

```diff
@@ -22,6 +30,193 @@
+        reduced_features_vector.append(
+            [feature_vector[index_row][index_column] for index, index_column in enumerate(index_list)])
+
+    return reduced_features_vector
+
+
+def get_map_to_index(X, features_indices, features_vector_size):
+    pass
+
+
+def to_dict(value_list_batch: List[list], feature_indices: Set[int] = None,
+            features_vector_size: int = ModelConfig.DEFAULT_FEATURES_VOCABULARY_SIZE):
+
+    if not feature_indices:
+        matrix = reduce_features_vector(value_list_batch, features_vector_size)
+
+        return [{index: value for index, value in enumerate(value_list)} for value_list in matrix]
+    else:
+        return [{index: value for index, value in enumerate(value_list) if index in feature_indices} for value_list
+                in value_list_batch]
+
+
+class FeaturesPreprocessor(BaseEstimator, TransformerMixin):
+    def __init__(self, features_indices: Iterable[int] = None,
+                 features_vocabulary_size=ModelConfig.DEFAULT_FEATURES_VOCABULARY_SIZE,
+                 features_map_to_index=None):
+        # feature_indices_set = None
+        if features_map_to_index is None:
+            features_map_to_index = []
+        self.features_vocabulary_size = features_vocabulary_size
+        self.features_indices = features_indices
+
+        # List of mappings to integers (corresponding to each feature column) of features values
+        # This value could be provided (model has been loaded) or not (first-time-training)
+        self.features_map_to_index = features_map_to_index
+
+    def fit(self, X):
+        if not self.features_indices:
+            indexes, mapping = reduce_features_to_indexes(X, self.features_vocabulary_size)
+        else:
+            indexes, mapping = reduce_features_to_indexes(X, self.features_vocabulary_size,
+                                                          indices=self.features_indices)
+
+        self.features_map_to_index = mapping
+        self.features_indices = indexes
+        return self
+
+    def transform(self, X, extend=False):
+        """
+        Transform the features into a vector, return the vector and the extracted number of features
+
+        :param extend: when set to true it's adding an additional empty feature list in the sequence.
+        """
+        features_vector = [[[self.features_map_to_index[index][value] if index in self.features_map_to_index and value in self.features_map_to_index[index] else 0
+                    for index, value in enumerate(value_list) if index in self.features_indices] for
+                    value_list in document] for document in X]
+
+        features_count = len(self.features_indices)
+
+        if extend:
+            for out in features_vector:
+                out.append([0] * features_count)
+
+        features_vector_padded, _ = pad_sequences(features_vector, [0] * features_count)
+        output = np.asarray(features_vector_padded)
+
+        return output, features_count
```

Comment:
I am just wondering whether there was a reason to return the `features_count`? This makes the the `transform` method a bit "unusual" in that it doesn't just return the transformed features.

Reply:
Right, the same information can be anyway accessed via ```len(preprocessor.features_indices)```